### PR TITLE
Create agx-client ip mapping on server side

### PR DESCRIPTION
**Pull request by obrask1**
Distribute the clients evenly on the available AGXs in an automated matter. The core intelligence behind the node distribution is a simple modulus operation:

`node_id = client_number % n_nodes` 

which I think is enough for now.

- [x] Script which creates a mapping between AGX ip address and client IDs based on global params
- [x] Give each client a unique random client ID
- [x] Include the AGX ip in the server messeage to the client
- [x] Communicate the client-agx mapping in a Yaml file to be able to visualize the distribution
- [x] Automatically update the docker-compose file from global params with correct ip addresses, agx nodes and deploy replicas.

### CODE REVIEW

File: src/fleetlearning/common/static_params.py [hunk 1]
```diff
@@ -27,12 +27,12 @@ def __init__(self):
         self.VM_KEY_PATH = "/home/nvidia/.ssh/id_rsa"
 
         self.DEVICE_DICT = {
-            "agx4.nodes.lab.ai.se": 0,
+            # "agx4.nodes.lab.ai.se": 0,
```
Comment (jonasandersson10) [inline src/fleetlearning/common/static_params.py:30]: should we default these to uncommented or commented in main?

File: src/fleetlearning/common/static_params.py [hunk 2]
```diff
@@ -62,9 +62,9 @@ def __init__(self):
         self.BATCH_SIZE = 8
         self.VAL_FACTOR = 0.1  # percentage of train data to use for validation
         self.SUBSET_FACTOR = 0.003  # subset of test frames to use
-        self.USE_GPU = True
-        self.NUM_GLOBAL_ROUNDS = 3
-        self.NUM_LOCAL_EPOCHS = 4
+        self.USE_GPU = False
```
Comment (jonasandersson10) [inline src/fleetlearning/common/static_params.py:65]: I think this should be defaulted to True, or? That's at least how the tests are right now. but we could change them if that makes more sense. 
```suggestion
        self.USE_GPU = True
```

File: src/fleetlearning/common/static_params.py [hunk 3]
```diff
@@ -62,9 +62,9 @@ def __init__(self):
...
+        self.NUM_GLOBAL_ROUNDS = 2
```
Comment (jonasandersson10) [inline src/fleetlearning/common/static_params.py:66]: change back?

File: src/fleetlearning/common/static_params.py [hunk 4]
```diff
@@ -62,9 +62,9 @@ def __init__(self):
...
+        self.NUM_LOCAL_EPOCHS = 1
```
Comment (jonasandersson10) [inline src/fleetlearning/common/static_params.py:67]: change back?

File: src/fleetlearning/docker-compose-template.yml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    command: python -u src/fleetlearning/client/client_main.py
+    depends_on:
+      - az-server
+      - agx-scheduler
+    deploy:
+      replicas: 2
+      restart_policy:
+        condition: "on-failure"
+    environment:
+      - "contraint:node==172.25.16.68"
```
Comment (jonasandersson10) [inline src/fleetlearning/docker-compose-template.yml:40]: Found this bug, might be why we saw weird behavior?
```suggestion
      - "constraint:node==172.25.16.68"
```

**Review by jonasandersson10 — COMMENTED** [review]
Great work @obrask1 ! I tried changing the number of clients in the docker-compose-template file and then run the `ip_utils.py` script. However, that didn't seem to change the vv-client entries in the final compose file. is something missing, or did I misunderstand something?

**Comment by jonasandersson10** [comment]
> Great work @obrask1 ! I tried changing the number of clients in the docker-compose-template file and then run the `ip_utils.py` script. However, that didn't seem to change the vv-client entries in the final compose file. is something missing, or did I misunderstand something?

I realized that I should've run the `configure_node_ips.py`. But  when I do that I get  `FileNotFoundError: [Errno 2] No such file or directory: 'src/fleetlearning/client_maps.yaml'`.

File: src/fleetlearning/common/ip_utils.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    node_to_client_map = dict.fromkeys(nodes, 0)
+    client_nr_to_client_id = {}
+
+    for client_number in range(1, n_clients + 1):
+        node_id = client_number % n_nodes if client_number % n_nodes != 0 else n_nodes
+        client_id = secrets.token_hex(id_bit_length)
+
+        client_nr_to_node_map[client_number] = nodes[node_id - 1]
+        client_id_to_node_map[client_id] = nodes[node_id - 1]
+        node_to_client_map[nodes[node_id - 1]] += 1
```
[the remark this answers is not in the document]
Comment (obrask1) [reply src/fleetlearning/common/ip_utils.py:30]: yes exactly!

File: src/fleetlearning/common/static_params.py (hunk 1 above)
Comment (obrask1) [reply src/fleetlearning/common/static_params.py:30]: good catch. I will reset all static_params to the same as current main

File: src/fleetlearning/common/static_params.py (hunk 2 above)
Comment (obrask1) [reply src/fleetlearning/common/static_params.py:65]: same here, I forgot to reset these!

File: src/fleetlearning/common/static_params.py (hunk 3 above)
Comment (obrask1) [reply src/fleetlearning/common/static_params.py:66]: same here, I forgot to reset these!

File: src/fleetlearning/docker-compose-template.yml (hunk 5 above)
Comment (obrask1) [reply src/fleetlearning/docker-compose-template.yml:40]: good catch!

**Comment by obrask1** [comment]
> Collaborator

My idea here was that we change everything in global params, then runs `configure_node_ips.py` to update the docker-compose. The server also runs this script in the background. Strange file error, from where are you running `configure_node_ips.py`? I think you need to be in root.

File: src/fleetlearning/docker-compose.yml [hunk 7]
```diff
@@ -1,40 +1,110 @@
...
-      - 59999:59999
+    - 59999:59999
     environment:
-      - "constraint:node==agx4.nodes.lab.ai.se"
+    - "constraint:node==agx4.nodes.lab.ai.se"
     deploy:
       replicas: 1
       restart_policy:
         condition: on-failure
+  agx-scheduler-2:
```
Comment (jonasandersson10) [inline src/fleetlearning/docker-compose.yml:29]: remove code below or keep all agx's? (this is autogenerated right?)

Comment (obrask1) [reply src/fleetlearning/docker-compose.yml:29]: yes, this is autogenerated based on the `DEVICE_DICT` in global_configs. I think it makes sense to keep all the settings in one place, so if a user only want some specific AGXes to be activated, they can simply be commented away there, and the client mapping script will not include them in docker-compose.

File: src/fleetlearning/common/ip_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (obrask1) [reply src/fleetlearning/common/ip_utils.py:30]: nice! done in new commit

File: src/fleetlearning/docker-compose.yml (hunk 7 above)
Comment (jonasandersson10) [reply src/fleetlearning/docker-compose.yml:29]: sounds perfect! 👍

**Review by jonasandersson10 — APPROVED** [review]
Awesome work @obrask1, let's merge! 🚀

**Pull request merged**